### PR TITLE
Avoid creating the bind group for indirect validation if buffer size is 0

### LINKED ISSUE #6414: Panic when indirect buffer is 0 sized
**Description**
I'm tracking `trunk` in my project and just hit a panic in changes related to https://github.com/gfx-rs/wgpu/pull/5714: 
```
thread 'test::unlit_textured_cube_material' panicked at /Users/schell/.cargo/git/checkouts/wgpu-53e70f8674b08dd4/59f56e0/wgpu-core/src/indirect_validation.rs:288:58:
called `Option::unwrap()` on a `None` value
```

I'm guessing it's this binding `binding_size` on line 276 in the file mentioned:
```
        let binding_size = calculate_src_buffer_binding_size(buffer_size, limits);
```

being used on line 288:
```
                size: Some(NonZeroU64::new(binding_size).unwrap()),
```

...so for some reason in my case `binding_size` is `0`. Maybe the buffer passed is `0`? 🤷 

My project creates buffers that are zero-sized as a placeholder until there's something to fill them with, at which point they'll be recreated. Is this a generally accepted pattern? If so, we probably need to handle that case?

Just to be clear, this is happening during buffer creation.

**Expected vs observed behavior**
I would expect to be able to create an indirect draw buffer that is zero sized. 

**Extra materials**
Backtrace:
```
thread 'test::unlit_textured_cube_material' panicked at /Users/schell/.cargo/git/checkouts/wgpu-53e70f8674b08dd4/59f56e0/wgpu-core/src/indirect_validation.rs:288:58:
called `Option::unwrap()` on a `None` value
stack backtrace:
   0: rust_begin_unwind
             at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/std/src/panicking.rs:665:5
   1: core::panicking::panic_fmt
             at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/core/src/panicking.rs:74:14
   2: core::panicking::panic
             at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/core/src/panicking.rs:148:5
   3: core::option::unwrap_failed
             at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/core/src/option.rs:2020:5
   4: core::option::Option<T>::unwrap
             at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/core/src/option.rs:970:21
   5: wgpu_core::indirect_validation::IndirectValidation::create_src_bind_group
             at /Users/schell/.cargo/git/checkouts/wgpu-53e70f8674b08dd4/59f56e0/wgpu-core/src/indirect_validation.rs:288:28
   6: wgpu_core::device::resource::Device::create_indirect_validation_bind_group
             at /Users/schell/.cargo/git/checkouts/wgpu-53e70f8674b08dd4/59f56e0/wgpu-core/src/device/resource.rs:780:30
   7: wgpu_core::device::resource::Device::create_buffer
             at /Users/schell/.cargo/git/checkouts/wgpu-53e70f8674b08dd4/59f56e0/wgpu-core/src/device/resource.rs:627:13
   8: wgpu_core::device::global::<impl wgpu_core::global::Global>::device_create_buffer
             at /Users/schell/.cargo/git/checkouts/wgpu-53e70f8674b08dd4/59f56e0/wgpu-core/src/device/global.rs:122:32
   9: <wgpu::backend::wgpu_core::ContextWgpuCore as wgpu::context::Context>::device_create_buffer
             at /Users/schell/.cargo/git/checkouts/wgpu-53e70f8674b08dd4/59f56e0/wgpu/src/backend/wgpu_core.rs:1196:13
  10: <T as wgpu::context::DynContext>::device_create_buffer
             at /Users/schell/.cargo/git/checkouts/wgpu-53e70f8674b08dd4/59f56e0/wgpu/src/context.rs:1583:20
  11: wgpu::api::device::Device::create_buffer
             at /Users/schell/.cargo/git/checkouts/wgpu-53e70f8674b08dd4/59f56e0/wgpu/src/api/device.rs:235:20
  12: <wgpu::api::buffer::Buffer as renderling::slab::cpu::IsBuffer>::buffer_create
             at ./src/slab/cpu.rs:283:9
  13: renderling::slab::cpu::SlabAllocator<Buffer>::recreate_buffer
             at ./src/slab/cpu.rs:438:35
  14: renderling::slab::cpu::SlabAllocator<Buffer>::upkeep
             at ./src/slab/cpu.rs:535:18
  15: renderling::draw::cpu::IndirectDraws::sync_with_internal_renderlets
             at ./src/draw/cpu.rs:81:45
  16: renderling::draw::cpu::DrawCalls::upkeep
             at ./src/draw/cpu.rs:278:13
  17: renderling::stage::cpu::Stage::tick_internal
             at ./src/stage/cpu.rs:701:9
  18: renderling::stage::cpu::Stage::render
             at ./src/stage/cpu.rs:734:27
  19: renderling::test::unlit_textured_cube_material
             at ./src/lib.rs:665:9
  20: renderling::test::unlit_textured_cube_material::{{closure}}
             at ./src/lib.rs:633:38
  21: core::ops::function::FnOnce::call_once
             at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/core/src/ops/function.rs:250:5
  22: core::ops::function::FnOnce::call_once
             at /rustc/eeb90cda1969383f56a2637cbd3037bdf598841c/library/core/src/ops/function.rs:250:5
```

### PR DESCRIPTION
**Pull request by teoxoy**
Fixes https://github.com/gfx-rs/wgpu/issues/6414.

### CODE REVIEW

File: tests/tests/dispatch_workgroups_indirect.rs [hunk 1]
```diff
@@ -68,18 +68,46 @@ static RESET_BIND_GROUPS: GpuTestConfiguration = GpuTestConfiguration::new()
...
+                "Indirect buffer uses bytes 0..12 which overruns indirect buffer of size 0",
+            )
+        }));
+    });
+
 async fn run_test(
     ctx: &TestingContext,
     num_workgroups: &[u32; 3],
     forget_to_set_bind_group: bool,
+    set_buffer_size_to_0: bool,
```
Comment (ErichDonGubler) [inline tests/tests/dispatch_workgroups_indirect.rs]: nitpick(non-blocking): It's starting to get really confusing which argument corresponds to what. I feel like we're at the threshold that a `struct` for the arguments is helpful.

Comment (ErichDonGubler) [inline tests/tests/dispatch_workgroups_indirect.rs]: nitpick: I don't see coverage where both `forget_to_set_bind_group` and `set_buffer_size_to_0` are `true`, but that sounds interesting to do. Maybe we could add this to the `RESET_BIND_GROUPS` test?

File: wgpu-core/src/indirect_validation.rs [hunk 2]
```diff
@@ -266,14 +266,19 @@ impl IndirectValidation {
...
         limits: &wgt::Limits,
         buffer_size: u64,
         buffer: &dyn hal::DynBuffer,
-    ) -> Result<Box<dyn hal::DynBindGroup>, DeviceError> {
+    ) -> Result<Option<Box<dyn hal::DynBindGroup>>, DeviceError> {
         let binding_size = calculate_src_buffer_binding_size(buffer_size, limits);
+        let binding_size = match NonZeroU64::new(binding_size) {
+            Some(binding_size) => binding_size,
+            None => return Ok(None),
+        };
```
Comment (ErichDonGubler) [inline wgpu-core/src/indirect_validation.rs]: style:

```suggestion
        let Some(binding_size) = NonZeroU64::new(binding_size) else {
            return Ok(None);
        };
```

File: tests/tests/dispatch_workgroups_indirect.rs (hunk 1 above)
Comment (ErichDonGubler) [reply tests/tests/dispatch_workgroups_indirect.rs]: ☝🏻Maybe this is because this function is trying to de-duplicate too much?

**Review by ErichDonGubler — APPROVED** [review]

File: tests/tests/dispatch_workgroups_indirect.rs (hunk 1 above)
Comment (teoxoy) [reply tests/tests/dispatch_workgroups_indirect.rs]: They each test validation individually though. A test where both would be true will have the same result as just `forget_to_set_bind_group` being true.

Comment (teoxoy) [reply tests/tests/dispatch_workgroups_indirect.rs]: Good point, I added a `TestResources` struct.

Comment (ErichDonGubler) [reply tests/tests/dispatch_workgroups_indirect.rs]: Maybe this is indicative that we should use a 3-state `enum` instead of two `bool`s, then? 🤔

Comment (teoxoy) [reply tests/tests/dispatch_workgroups_indirect.rs]: This is no longer an issue with the refactored test file.

**Pull request merged**